experiments/2017-09-21-rust-on-android.md

```diff
@@ -0,0 +1,224 @@
+
+```
+public class RustGreetings {
+
+    private static native String greeting(final String pattern);
+
+    public String sayHello(String to) {
+        return greeting(to);
+    }
+}
+```
+
+Instead of creating  C header that will be used as a bridge as we did when we wanted to integrate with Swift, for Android we want to expose our functions through JNI. The way that JNI constructs the name of the function that it will call is `Java_<domain>_<class>_<methodname>`. In the case of the method, `greeting` that we have declared here, the function in our Rust library that JNI will attempt to call will be `Java_com_mozilla_greetings_RustGreetings_greeting`. This is the reason why we created our Android project and Java wrapper class before adding any JNI code to the rust library. We needed to know what the domain, class and function name were before we could construct the right JNI function name in Rust. Let's head back over to our Rust project and create the partner code.
+
+Open `cargo/src/lib.rs`. At the bottom of the file add the following code:
+```
+/// Expose the JNI interface for android below
+#[cfg(target_os="android")]
+#[allow(non_snake_case)]
+pub mod android {
+    extern crate jni;
+
+    use super::*;
+    use self::jni::JNIEnv;
+    use self::jni::objects::{JClass, JString};
+    use self::jni::sys::{jstring};
+
+    #[no_mangle]
+    pub unsafe extern fn Java_com_mozilla_greetings_RustGreetings_greeting(env: JNIEnv, _: JClass, java_pattern: JString) -> jstring {
+        let world = rust_greeting(env.get_string(java_pattern).expect("invalid pattern string").as_ptr());
+        let output = env.new_string(CStr::from_ptr(world).to_str().unwrap()).expect("Couldn't create java string!");
+        rust_greeting_free(world);
+
+        output.into_inner()
+    }
+}
+```
+The first line here `#[cfg(target_os="android")]` is telling the compiler to target Android when compiling this module. `#[cfg]` is a special attribute that allows you to compile code based on a flag passed to the compiler. 
+
+The second line, `#[allow(non_snake_case)]`, tells the compiler not to warn if we are not using snake_case for a variable or function name. The Rust compiler is very strict, this is one of the things that makes Rust great - it's nearly impossible to code something that is not correct, and it enforces the use of snake_case throughout. However, we defined our class name and native method in our Android project using Java coding conventions which is camelCase and UpperCamelCase and we don't want to change this or our Java code will look wrong. Given the way that JNI constructs native function names, we need to tell the Rust compiler to go easy on us in this instance. This flag will apply to all functions and variables created inside this module that we are creating, called `android`. 
```

Comment:
em dash (—) instead of hyphen (-).

Perhaps use `camelCase` etc. instead of camelCase.